### PR TITLE
CarpetX: Change the implementation to Driver.

### PR DESCRIPTION
**Pull request by rhaas80**
~~This involves adding a parameter "periodic" whose only allowed value is "no."~~
This involves adding a parameter "periodic" that must be set to "yes" for any periodicity to be active.
 It also involves modifying parameter files, file names, and source files ~~where the name "CarpetX" needs to be replaced by "Driver."~~ to account for the fact that the `regrid_error` grid function moved from `CarpetX` to `CarpetRegridX`.
 
 Parameter files may, but not have to use `Driver::periodic` instead of `CarpetX::periodic` since the parameter file parser accepts both implementation and thorn names for restricted parameters.
 
 ```
ActiveThorns = PUGH

PUGH::periodic = yes
Driver::periodic_x = no
 ```
This picks up the work in https://github.com/EinsteinToolkit/CarpetX/pull/299

### CODE REVIEW

**Comment by rhaas80** [comment]
Ticket is here https://bitbucket.org/einsteintoolkit/tickets/issues/2813/allow-carpetx-to-implement-driver

File: TestSymmetries/par/test_symmetry_xperiodic_yperiodic_zperiodic.par [hunk 1]
```diff
@@ -19,9 +19,9 @@ CarpetX::ncells_x = 8
 CarpetX::ncells_y = 8
 CarpetX::ncells_z = 8
 
-CarpetX::periodic_x = yes
-CarpetX::periodic_y = yes
-CarpetX::periodic_z = yes
+Driver::periodic_x = yes
```
Comment (eschnett) [inline TestSymmetries/par/test_symmetry_xperiodic_yperiodic_zperiodic.par:22]: This seems wrong; `Driver::periodic` is not set.

File: WaveToyX/par/standing.par [hunk 2]
```diff
@@ -13,9 +13,9 @@ Cactus::presync_mode = "mixed-error"
 
 CarpetX::poison_undefined_values = no
 
-CarpetX::periodic_x = yes
-CarpetX::periodic_y = yes
-CarpetX::periodic_z = yes
+Driver::periodic_x = yes
```
Comment (eschnett) [inline WaveToyX/par/standing.par:16]: This seems wrong; `Driver::periodic` is not set.

File: WaveToyX/test/presync.par [hunk 3]
```diff
@@ -35,7 +36,6 @@ CarpetX::out_norm_omit_sumloc_for_backward_compatibility = yes
 
 CarpetX::out_tsv_vars = "
     WaveToyX::state
-    WaveToyX::rhs
```
Comment (eschnett) [inline WaveToyX/test/presync.par:38]: This change seems unrelated – why is the test weakened?

File: CarpetX/param.ccl [hunk 4]
```diff
@@ -69,18 +69,23 @@ CCTK_INT ncells_z "Number of grid cells"
 
 
 
-BOOLEAN periodic_x "Periodic"
+RESTRICTED:
+
+BOOLEAN periodic "Periodic boundary conditions"
```
Comment (eschnett) [inline CarpetX/param.ccl:74]: The comments on these parameters should be improved. They should explain under which conditions there is a periodic boundary in a particular direction.

File: CarpetXRegrid/README [hunk 5]
```diff
@@ -0,0 +1,11 @@
+Cactus Code Thorn Regrid
```
Comment (eschnett) [inline CarpetXRegrid/README:1]: The thorn name is wrong

File: CarpetXRegrid/README [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+Author(s)    : Steven R. Brandt <sbrandt@cct.lsu.edu>
+Maintainer(s): Steven R. Brandt <sbrandt@cct.lsu.edu>
+Licence      : Cactus
```
Comment (eschnett) [inline CarpetXRegrid/README:4]: That's not a proper licence name

File: CarpetXRegrid/configuration.ccl [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# Configuration definitions for thorn Regrid
```
Comment (eschnett) [inline CarpetXRegrid/configuration.ccl:1]: The thorn name is wrong

File: CarpetXRegrid/configuration.ccl [hunk 8]
```diff
@@ -0,0 +1,6 @@
...
+REQUIRES AMReX Arith Loop 
+
+PROVIDES Regrid
```
Comment (eschnett) [inline CarpetXRegrid/configuration.ccl:4]: The capability should probably be called `CarpetXRegrid` instead. It's not a standardized capability in the Einstein Toolkit.

File: CarpetXRegrid/interface.ccl [hunk 9]
```diff
@@ -0,0 +1,9 @@
+# Interface definition for thorn Regrid
```
Comment (eschnett) [inline CarpetXRegrid/interface.ccl:1]: The thorn name is wrong

File: CarpetXRegrid/interface.ccl [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+implements: CarpetXRegrid
```
Comment (eschnett) [inline CarpetXRegrid/interface.ccl:2]: We usually use all caps keywords, i.e. `IMPLEMENTS`.

They keyword `inherits` is not necessary since nothing is inherited.

File: CarpetXRegrid/param.ccl [hunk 11]
```diff
@@ -0,0 +1 @@
+# Parameter definitions for thorn Regrid
```
Comment (eschnett) [inline CarpetXRegrid/param.ccl:1]: Thorn name is wrong

File: CarpetXRegrid/schedule.ccl [hunk 12]
```diff
@@ -0,0 +1,6 @@
+# Schedule definitions for thorn Regrid
```
Comment (eschnett) [inline CarpetXRegrid/schedule.ccl:1]: Thorn name is wrong

File: CarpetXRegrid/src/make.code.defn [hunk 13]
```diff
@@ -0,0 +1,7 @@
+# Main make.code.defn file for thorn Regrid
```
Comment (eschnett) [inline CarpetXRegrid/src/make.code.defn:1]: The thorn name is wrong

File: CarpetXRegrid/src/regrid_error.cxx [hunk 14]
```diff
@@ -1,18 +1,17 @@
...
 
 namespace CarpetX {
+using namespace std;
 
 extern "C" void CarpetX_InitError(CCTK_ARGUMENTS) {
   DECLARE_CCTK_ARGUMENTSX_CarpetX_InitError;
   DECLARE_CCTK_PARAMETERS;
 
-  grid.loop_device<1, 1, 1, where_t::everywhere>(
+  grid.loop_device<1, 1, 1, Loop::where_t::everywhere>(
```
Comment (eschnett) [inline CarpetXRegrid/src/regrid_error.cxx:14]: Maybe use the simpler `loop_all_device` instead?

File: FluxWaveToyX/par/standing.par [hunk 15]
```diff
@@ -13,9 +13,9 @@ Cactus::presync_mode = "mixed-error"
 
 CarpetX::poison_undefined_values = no
 
-CarpetX::periodic_x = yes
-CarpetX::periodic_y = yes
-CarpetX::periodic_z = yes
+Driver::periodic_x = yes
```
Comment (eschnett) [inline FluxWaveToyX/par/standing.par:16]: This doesn't enable periodic boundaries, does it?

File: SpacetimeWaveToyX/test/standing/standing.par [hunk 16]
```diff
@@ -15,9 +15,9 @@ CarpetX::ncells_x = 8
 CarpetX::ncells_y = 8
 CarpetX::ncells_z = 8
 
-CarpetX::periodic_x = yes
-CarpetX::periodic_y = yes
-CarpetX::periodic_z = yes
+Driver::periodic_x = yes
```
Comment (eschnett) [inline SpacetimeWaveToyX/test/standing/standing.par:18]: This doesn't enable periodic boundaries.

File: StaggeredWaveToyX/test/standing/standing.par [hunk 17]
```diff
@@ -16,9 +16,9 @@ CarpetX::ncells_x = 8
 CarpetX::ncells_y = 8
 CarpetX::ncells_z = 8
 
-CarpetX::periodic_x = yes
-CarpetX::periodic_y = yes
-CarpetX::periodic_z = yes
+Driver::periodic_x = yes
```
Comment (eschnett) [inline StaggeredWaveToyX/test/standing/standing.par:19]: This doesn't enable periodic boundaries.

File: TestODESolvers/test/test-rkf78.par [hunk 18]
```diff
@@ -17,6 +17,7 @@ CarpetX::blocking_factor_z = 1
 
 CarpetX::ghost_size = 0
 
+Driver::periodic = "yes"
 CarpetX::periodic_x = yes
```
Comment (eschnett) [inline TestODESolvers/test/test-rkf78.par:21]: Shouldn't these parameters be renamed to `Driver::periodic_x` etc.?

**Review by eschnett — CHANGES_REQUESTED** [review]

File: WaveToyX/test/presync.par (hunk 3 above)
Comment (rhaas80) [reply WaveToyX/test/presync.par:38]: good question, no idea why. I'll put it back.

File: CarpetXRegrid/README (hunk 5 above)
Comment (rhaas80) [reply CarpetXRegrid/README:1]: @stevenrbrandt

File: CarpetX/param.ccl (hunk 4 above)
Comment (rhaas80) [reply CarpetX/param.ccl:74]: will do.

File: WaveToyX/par/standing.par (hunk 2 above)
Comment (rhaas80) [reply WaveToyX/par/standing.par:16]: yes, this should be `Driver::periodic = yes`. Hmm, I thought the tests had all passed. Will check.

File: CarpetXRegrid/README (hunk 6 above)
Comment (rhaas80) [reply CarpetXRegrid/README:4]: True. @stevenrbrandt best might be to pick the same license that CarpetX uses.

File: CarpetXRegrid/configuration.ccl (hunk 7 above)
Comment (rhaas80) [reply CarpetXRegrid/configuration.ccl:1]: @stevenrbrandt

File: CarpetXRegrid/configuration.ccl (hunk 8 above)
Comment (rhaas80) [reply CarpetXRegrid/configuration.ccl:4]: @stevenrbrandt

File: CarpetXRegrid/interface.ccl (hunk 9 above)
Comment (rhaas80) [reply CarpetXRegrid/interface.ccl:1]: @stevenrbrandt

File: CarpetXRegrid/interface.ccl (hunk 10 above)
Comment (rhaas80) [reply CarpetXRegrid/interface.ccl:2]: will do

File: CarpetXRegrid/schedule.ccl (hunk 12 above)
Comment (rhaas80) [reply CarpetXRegrid/schedule.ccl:1]: @stevenrbrandt

File: CarpetXRegrid/src/make.code.defn (hunk 13 above)
Comment (rhaas80) [reply CarpetXRegrid/src/make.code.defn:1]: @stevenrbrandt

File: FluxWaveToyX/par/standing.par (hunk 15 above)
Comment (rhaas80) [reply FluxWaveToyX/par/standing.par:16]: no it does not, this should be `Driver::periodic = yes`. Hmm, I thought the tests had all passed. Will check.

File: SpacetimeWaveToyX/par/standing.par [hunk 19]
```diff
@@ -12,9 +12,9 @@ Cactus::presync_mode = "mixed-error"
 
 CarpetX::poison_undefined_values = no
 
-CarpetX::periodic_x = yes
-CarpetX::periodic_y = yes
-CarpetX::periodic_z = yes
+Driver::periodic_x = yes
```
[the remark this answers is not in the document]
Comment (rhaas80) [reply SpacetimeWaveToyX/par/standing.par:15]: no it does not, this should be `Driver::periodic = yes`. Hmm, I thought the tests had all passed. Will check.

File: SpacetimeWaveToyX/test/standing/standing.par (hunk 16 above)
Comment (rhaas80) [reply SpacetimeWaveToyX/test/standing/standing.par:18]: no it does not, this should be Driver::periodic = yes. Hmm, I thought the tests had all passed. Will check.

File: StaggeredWaveToyX/par/standing.par [hunk 20]
```diff
@@ -14,9 +14,9 @@ Cactus::presync_mode = "mixed-error"
 CarpetX::verbose = no
 CarpetX::poison_undefined_values = no
 
-CarpetX::periodic_x = yes
-CarpetX::periodic_y = yes
-CarpetX::periodic_z = yes
+Driver::periodic_x = yes
```
[the remark this answers is not in the document]
Comment (rhaas80) [reply StaggeredWaveToyX/par/standing.par:17]: no it does not, this should be Driver::periodic = yes. Hmm, I thought the tests had all passed. Will check.

File: StaggeredWaveToyX/test/standing/standing.par (hunk 17 above)
Comment (rhaas80) [reply StaggeredWaveToyX/test/standing/standing.par:19]: no it does not, this should be Driver::periodic = yes. Hmm, I thought the tests had all passed. Will check.

File: TestODESolvers/test/test-rkf78.par (hunk 18 above)
Comment (rhaas80) [reply TestODESolvers/test/test-rkf78.par:21]: should be , yes.

File: TestSymmetries/par/test_symmetry_xperiodic_yperiodic_zperiodic.par (hunk 1 above)
Comment (rhaas80) [reply TestSymmetries/par/test_symmetry_xperiodic_yperiodic_zperiodic.par:22]: yes, this should be `Driver::periodic = yes`. Hmm, I thought the tests had all passed. Will check.

**Comment by rhaas80** [comment]
Will need to add CarpetXRegrid to the thornlist used by the CI testing.

File: WaveToyX/par/standing.par (hunk 2 above)
Comment (rhaas80) [reply WaveToyX/par/standing.par:17]: update, is not a test so nothing failed

File: WaveToyX/test/presync.par (hunk 3 above)
Comment (rhaas80) [reply WaveToyX/test/presync.par:38]: updated, is not a test so nothing failed

Comment (rhaas80) [reply WaveToyX/test/presync.par:38]: ah, the output file (`WaveToyX/test/presync/wavetoyx-rhs.it000000.x.tsv` or so) was never committed to git, so there was no actual test. See 89f1a58d

File: CarpetXRegrid/src/regrid_error.cxx (hunk 14 above)
Comment (rhaas80) [reply CarpetXRegrid/src/regrid_error.cxx]: done

File: SIMDWaveToyX/par/standing.par [hunk 21]
```diff
@@ -13,9 +13,9 @@ Cactus::presync_mode = "mixed-error"
 
 CarpetX::poison_undefined_values = no
 
-CarpetX::periodic_x = yes
-CarpetX::periodic_y = yes
-CarpetX::periodic_z = yes
+Driver::periodic_x = yes
```
[the remark this answers is not in the document]
Comment (rhaas80) [reply SIMDWaveToyX/par/standing.par:17]: was not a test so did not fail. Fixed now.

File: SpacetimeWaveToyX/test/standing/standing.par (hunk 16 above)
Comment (rhaas80) [reply SpacetimeWaveToyX/test/standing/standing.par:18]: That's in the output directory. The par file should not even be there. Removed.

File: StaggeredWaveToyX/test/standing/standing.par (hunk 17 above)
Comment (rhaas80) [reply StaggeredWaveToyX/test/standing/standing.par:19]: That's in the output directory. The par file should not even be there. Removed.

File: TestODESolvers/test/test-rkf78.par (hunk 18 above)
Comment (rhaas80) [reply TestODESolvers/test/test-rkf78.par]: That's in the output directory. The par file should not even be there. Removed.

File: StaggeredWaveToyX/par/standing.par (hunk 20 above)
[the remark this answers is not in the document]
Comment (rhaas80) [reply StaggeredWaveToyX/par/standing.par:18]: That's in the output directory. The par file should not even be there. Removed.

File: SpacetimeWaveToyX/par/standing.par (hunk 19 above)
[the remark this answers is not in the document]
Comment (rhaas80) [reply SpacetimeWaveToyX/par/standing.par:16]: was not a test so did not fail. Fixed now.

File: FluxWaveToyX/par/standing.par (hunk 15 above)
Comment (rhaas80) [reply FluxWaveToyX/par/standing.par:17]: was not a test so did not fail. Fixed now.

File: CarpetX/param.ccl (hunk 4 above)
Comment (rhaas80) [reply CarpetX/param.ccl:74]: done.

**Comment by eschnett** [comment]
"This involves adding a parameter "periodic" whose only allowed value is "no."". is that still true?

File: BoxInBox/test/levels1.par [hunk 22]
```diff
@@ -3,6 +3,7 @@ ActiveThorns = "
     CarpetX
     CoordinatesX
     IOUtil
+    CarpetXRegrid
```
Comment (eschnett) [inline BoxInBox/test/levels1.par:6]: Can you sort thorn names alphabetically?

File: CarpetX/interface.ccl [hunk 23]
```diff
@@ -1,8 +1,6 @@
 # Interface definition for thorn CarpetX
 
-IMPLEMENTS: CarpetX
-
-INHERITS: IO
+IMPLEMENTS: Driver
```
Comment (eschnett) [inline CarpetX/interface.ccl:3]: That's it! That's the core of the change! Yay!

**Comment by rhaas80** [comment]
> "This involves adding a parameter "periodic" whose only allowed value is "no."". is that still true?

not true anymore. It does introduce a parameter `periodic` but it must be set to "yes" for any periodicity to be applied.

File: CarpetXRegrid/schedule.ccl [hunk 24]
```diff
@@ -0,0 +1,6 @@
+# Schedule definitions for thorn CarpetXRegrid
+SCHEDULE CarpetX_InitError AT basegrid
```
Comment (eschnett) [inline CarpetXRegrid/schedule.ccl:2]: The function name prefix should be `CarpetXRegrid`.

File: StaggeredWaveToyX/test/standing/standing.par (hunk 17 above)
Comment (eschnett) [inline StaggeredWaveToyX/test/standing/standing.par:19]: Is a `Driver::periodic = yes` missing here?

**Comment by rhaas80** [comment]
> > "This involves adding a parameter "periodic" whose only allowed value is "no."". is that still true?
> 
> not true anymore. It does introduce a parameter `periodic` but it must be set to "yes" for any periodicity to be applied.

Turns out that (see updated description) the "adapt to CarpetX implementation change" commits aren't actually required since the parameter file parser will accept both implementation and thorn names for restricted parameters (and private ones as well, which is a bug...).

**Comment by rhaas80** [comment]
> There should be checkpoint/recovery tests as well. I think the respective checkpoint files need to be regenerated.

The TestOuptut/recover-openpmd test passes:

```
  [10] recover-openpmd
       CarpetX/TestOutput/test/recover-openpmd.par
  Choose test: [0] 
   --> 10

------------------------------------------------------------------------

  Test TestOutput: recover-openpmd
    "CarpetX/TestOutput/test/recover-openpmd.par"

  Issuing mpirun -np 2 /data/rhaas/postdoc/gr/cactus/ET_trunk/exe/cactus_sim /data/rhaas/postdoc/gr/cactus/ET_trunk/arrangements/CarpetX/TestOutput/test/recover-openpmd.par

  Success: 40 files identical
```

Looking at file `md.0` it only has settings for parameters:

```
CarpetX::periodic_x = no
CarpetX::periodic_y = no
CarpetX::periodic_z = no
```

which are apparently acceptable (not sure why, http://einsteintoolkit.org/referencemanual/ReferenceManual.html#x1-156000A2 clearly says it needs to be the implementation name "Driver").

The missing `Driver::periodic` is not caught because it is a new parameter.

File: SpacetimeWaveToyX/test/standing/standing.par (hunk 16 above)
[the remark this answers is not in the document]
Comment (rhaas80) [reply SpacetimeWaveToyX/test/standing/standing.par:18]: That's in the output directory. The parfile should be deleted there.

File: StaggeredWaveToyX/test/standing/standing.par (hunk 17 above)
Comment (rhaas80) [reply StaggeredWaveToyX/test/standing/standing.par:19]: That's in the output directory. The parfile should be deleted there.

File: BoxInBox/test/levels1.par (hunk 22 above)
Comment (rhaas80) [reply BoxInBox/test/levels1.par:6]: yes

File: CarpetX/configuration.ccl [hunk 25]
```diff
@@ -2,7 +2,7 @@
 
 REQUIRES AMReX IOUtil MPI yaml_cpp zlib
 
-REQUIRES Arith Loop
+REQUIRES Arith Loop CarpetXRegrid
```
[the remark this answers is not in the document]
Comment (rhaas80) [reply CarpetX/configuration.ccl:5]: yes

**Comment by eschnett** [comment]
FYI: I'm looking into why the OneAPI tests fail in a separate PR.

**Comment by eschnett** [comment]
The OneAPI build is fixed; try merging in the `main` branch.

**Comment by rhaas80** [comment]
I rebased car[etx-driver onto main (I tend to prefer linear history over merges). Note that one cannot merge anything into main unless the change request is removed.

**Review by eschnett — APPROVED** [review]
